### PR TITLE
feat: added user registration functionality:

### PR DESCRIPTION
**Pull request by PavloMelnyk10**
- User entity with required fields
- User repository for database interaction
- User mapper for entity-DTO conversion
- Registration and response DTOs
- Data validation for registration
- FieldMatch annotation for password validation
- Registration endpoint in AuthenticationController
- GlobalExceptionHandler for RegistrationException
- UserAlreadyExistsException for duplicate registration handling

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/controller/AuthController.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import mate.academy.bookstore.dto.user.UserResponseDto;
+import mate.academy.bookstore.exception.RegistrationException;
+import mate.academy.bookstore.service.user.UserService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/auth")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthController.java]: add `server.servlet.context-path=/api` to application.properties

```suggestion
@RequestMapping("/auth")
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
@Getter
@Setter
```

File: src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new UserAlreadyExistsException("Can't register user. Email already exists");
+        }
+
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java]: use MapStruct to map user

File: src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
+        User savedUser = userRepository.save(user);
+
+        return userMapper.toUserResponse(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);

        return userMapper.toUserResponse(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/controller/AuthController.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/auth")
+@RequiredArgsConstructor
+public class AuthController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthController.java:23]: add Swagger annotations

File: src/main/java/mate/academy/bookstore/controller/BookController.java [hunk 6]
```diff
@@ -25,7 +25,7 @@
 @Tag(name = "Book management", description = "Endpoints for managing books")
 @RequiredArgsConstructor
 @RestController
-@RequestMapping(value = "/api/books")
+@RequestMapping(value = "/books")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/BookController.java]:
```suggestion
@RequestMapping("/books")
```

File: src/main/java/mate/academy/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -60,6 +60,28 @@ public ResponseEntity<Object> handleEntityNotFound(
         return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException e) {
+        Map<String, Object> body = createErrorBody(
+                HttpStatus.BAD_REQUEST.value(),
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
                HttpStatus.CONFLICT.value(),
```

File: src/main/java/mate/academy/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 8]
```diff
@@ -60,6 +60,28 @@ public ResponseEntity<Object> handleEntityNotFound(
...
+                "Registration Error",
+                List.of(e.getMessage())
+        );
+
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        return new ResponseEntity<>(body, HttpStatus.CONFLICT);
```

File: src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new UserAlreadyExistsException("Can't register user. Email already exists");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java]:
```suggestion
            throw new RegistrationException("Can't register user. Email already exists");
```

File: src/main/java/mate/academy/bookstore/exception/UserAlreadyExistsException.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.bookstore.exception;
+
+public class UserAlreadyExistsException extends RuntimeException {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/exception/UserAlreadyExistsException.java]: remove and use RegistrationException

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Getter
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted = false")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]: Check if `@Where` is deprecated

File: src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java (hunk 4 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java]: fix comment

File: src/main/java/mate/academy/bookstore/validation/FieldMatchValidator.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        this.firstFieldName = constraintAnnotation.first();
+        this.secondFieldName = constraintAnnotation.second();
+        this.message = constraintAnnotation.message();
+    }
+
+    @Override
+    public boolean isValid(UserRegistrationRequestDto value, ConstraintValidatorContext context) {
+        if (value == null) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/validation/FieldMatchValidator.java:24]: It's better to simplify 

See example
```suggestion
        Object field = new BeanWrapperImpl(value).getPropertyValue(this.field);
        Object fieldMatch = new BeanWrapperImpl(value).getPropertyValue(this.fieldMatch);
        return Objects.equals(field, fieldMatch);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 13]
```diff
@@ -9,15 +9,19 @@
...
 
 @Entity
 @Getter
 @Setter
+@FilterDef(name = "isDeletedFilter",
+        parameters = @ParamDef(name = "isDeleted", type = Boolean.class))
 @Table(name = "books")
 @SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
-@Where(clause = "is_deleted=false")
+@Filter(name = "isDeletedFilter", condition = "is_deleted = false")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java:24]:
```suggestion
@SQLRestriction(name = "isDeletedFilter", condition = "is_deleted = false")
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.Filter;
+import org.hibernate.annotations.SQLDelete;
+
+@Getter
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@Filter(name = "isDeletedFilter", condition = "is_deleted = false")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java:18]:
```suggestion
@SQLRestriction(name = "isDeletedFilter", condition = "is_deleted = false")
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Column(nullable = false)
+    private String firstName;
+
+    @Column(nullable = false)
+    private String lastName;
+
+    private String shippingAddress;
+
+    @Column(name = "is_deleted")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(requestDto.getEmail())) {
```

File: src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+            throw new RegistrationException("Can't register user. Email already exists");
+        }
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/user/UserServiceImpl.java]: remove redundant empty lines

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/model/Book.java (hunk 13 above)
Comment (PavloMelnyk10) [reply src/main/java/mate/academy/bookstore/model/Book.java:24]: It doesn't work with @SQLRestriction. It's recommended to use @Filter. [https://www.baeldung.com/spring-jpa-soft-delete](url).

Comment (Elena-Bruyako) [reply src/main/java/mate/academy/bookstore/model/Book.java:24]: got it

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**